edk II C Coding Standard: File and directory naming guidelines

changab opened the pull request:
Add file and directory naming guidelines for EDKII modules. Also consider the processor architecture and vendor implementation.

This is the draft version to community for the wide discussion.

Signed-off-by: Abner Chang <abner.chang@amd.com>
Cc: Ray Ni <ray.ni@intel.com>
Cc: Michael D Kinney <michael.d.kinney@intel.com>
Cc: Sunil V L <sunilvl@ventanamicro.com>
Cc: Abdul Lateef Attar <abdattar@amd.com>

niruiyu commented on 4_naming_conventions/42_file_names.md:
````diff
@@ -58,3 +58,127 @@ regular expression:
+Example:
+   CpuIo2Dxe.inf
+```
+
````
The above is not necessary because it's a special case of the following rule.

changab replied:
I think the above one says if the module is generic to all CPU archs or >=2 CPU archs (X64 and IA32)? @mdkinney 
The following one should be changed to "If the implementation is considered as only for the specific CPU arch or vendor. 
<CpuArch> should be REQUIRED, right?

mdkinney replied:
We have to be careful here.  Just because the very first implementation of a component may be for a single CPU or CPU+Vendor, that does not mean that additional CPU/Vendor support may not be added in the future.  We do not want to change the directory of INF name later because that breaks backwards compatibility.  The rule below must only be used if the developer can not imagine any other CPU/Vendor combinations every being added later

niruiyu replied:
I still think this rule is not needed.
When the optional CpuArch and Vendor are absent in the file name, following rule can clearly describe this rule.

I want to simplify the rule so that people can remember/understand and follow that easily.

changab replied:
We probably can mention "If the implementation supports >=2 or all CPU archs" case on the wiki page. That is ok we don't assign <arch> or <vendor> when the first time to implement the module. We won't rename the directory or INF. We will create a new INF file for arch or vendor instead.
Yeah, seems to me <ModuleName><Phase><CpuArch><Vendor>.inf covers <ModuleName><Phase>.inf case.

niruiyu commented on 4_naming_conventions/42_file_names.md:
````diff
@@ -58,3 +58,127 @@ regular expression:
+   CpuIo2DxeAmd.inf
+```
+
+If the implementation does not have any shared code between phases. The module INF is located under \<Feature\>/\<Phase\> or \<Feature\>/<Phase\>/<CpuArch\>/<Vendor\> (see section 4.6 Directory Names).
````
In MdeModulePkg, PCD driver follows this rule. But ResetSystemRuntimeDxe and ResetSystemPei don't follow.
Can we remove this rule so that the rule is clearer.

changab replied:
Or the spec is only applied to the newly introduced module or reconstructions after the doc is published?

mdkinney replied:
Both styles are allowed.  It is developer choice.

niruiyu commented on 4_naming_conventions/42_file_names.md:
```diff
@@ -58,3 +58,127 @@ regular expression:
 
 That is, a letter followed by zero, or more, letters, underscores, dashes, or
 digits followed by a period followed by one or more letters or digits.
+
```
How about put the directory naming rule before the file naming rule?

changab replied:
This is related to the chapter structure in this spec. 4.2 is mainly for the file, so add Directory Names in 4.6.

changab replied:
What are the impacts to insert a section in between?

mdkinney replied:
No issues with inserting sections within.  Just be aware if you reorder items that are linked in Summary.md, you have to update that file too

niruiyu commented on 4_naming_conventions/42_file_names.md:
```diff
@@ -58,3 +58,127 @@ regular expression:
+
+If the implementation does not have any shared code between phases. The module INF is located under \<Feature\>/\<Phase\> or \<Feature\>/<Phase\>/<CpuArch\>/<Vendor\> (see section 4.6 Directory Names).
+
+* If the implementation supports >=2 or all CPU archs:
```
I do not understand the following rules.

niruiyu commented on 4_naming_conventions/42_file_names.md:
````diff
@@ -58,3 +58,127 @@ regular expression:
+```
+#### 4.2.5.3 EDK II INF file within a Library instance
+```
+<Phase><CpuArch><Vendor><LibraryClassName><Dependency>.inf
````
Good to me.

niruiyu commented on 4_naming_conventions/42_file_names.md:
````diff
@@ -58,3 +58,127 @@ regular expression:
+#### 4.2.5.4 EDK II source files within a Library/Module instance
+
+```
+<FileName>.c
````
shall it be moved to directory chapter?

changab replied:
This seems the files and directory mixed.

changab replied:
This section mainly talks about the source file.

niruiyu commented on 4_naming_conventions/46_directory_names.md:
```diff
@@ -0,0 +1,114 @@
+
+-->
+
+## 4.6 Directory Names
```
again, can we move this chapter before file naming chapter?

changab replied:
What are the impacts to insert a section in between?

changab replied:
yes. I can move directory names to prior file names

niruiyu commented on 4_naming_conventions/46_directory_names.md:
````diff
@@ -0,0 +1,114 @@
+
+* If the implementation supports >=2 or all CPU archs:
+```
+<ModuleName><Phase>
````
Shall we say \<Feature\>\<Phase\> so that not introducing another \<ModuleName\> keyword?

changab replied:
need more details of this question

niruiyu replied:
Update the question. This comment is similar to the one in below.
Usually ModuleName is short name of a Feature. E.g.: PCD(Dxe), Variable(RuntimeDxe), CpuMp(Pei).
Think about how a developer chooses a ModuleName when he creates a new module.

changab replied:
Hmm, seems like we can replace all \<ModuleName\> with \<Feature\>. @mdkinney ?

niruiyu commented on 4_naming_conventions/46_directory_names.md:
````diff
@@ -0,0 +1,114 @@
+
+* If the implementation is for the specific CPU arch or vendor:
+```
+<ModuleName><Phase><CpuArch><Vendor>
````
good to me except replace ModuleName with Feature

changab replied:
The <modulename> is not the same as the <feature> in the usual case?

niruiyu replied:
If you replace ModuleName with Feature, this rule is consistent to the rule defined for file names.

niruiyu commented on 4_naming_conventions/46_directory_names.md:
````diff
@@ -0,0 +1,114 @@
+```
+
+* If the implementation does not have any shared code between phases (e.g.
+MdeModulePkg/Universal/PCD) and supports >=2 or all CPU archs:
````
With the replacement in above rule, this rule is not needed.

niruiyu commented on 4_naming_conventions/46_directory_names.md:
````diff
@@ -0,0 +1,114 @@
+
+#### 4.6.2 EDKII Library directory
+```
+<Phase><CpuArch><Vendor><LibraryClassName><Dependency>
````
Good to me.

vlsunil commented on 4_naming_conventions/46_directory_names.md:
```diff
@@ -0,0 +1,114 @@
+                                       that the implementation is layered on top of).
+
+Example:
+   SmmAmdSmmCpuFeaturesLib/
```
Wondering whether we can have better example. Smm comes twice.

changab committed:
edk II C Coding Standard: Insert Directory Names section

Insert 4.2 Directory names as the placeholder and update
markdown file names and content of follow up sections.

PR: https://github.com/tianocore-docs/edk2-CCodingStandardsSpecification/pull/2/files

Signed-off-by: Abner Chang <abner.chang@amd.com>
Cc: Ray Ni <ray.ni@intel.com>
Cc: Michael D Kinney <michael.d.kinney@intel.com>
Cc: Sunil V L <sunilvl@ventanamicro.com>
Cc: Abdul Lateef Attar <abdattar@amd.com>
Cc: Leif Lindholm <quic_llindhol@quicinc.com>
Reviewed-by: Ray Ni <Ray.ni@intel.com>
Reviewed-by: Abdul Lateef Attar <abdattar@amd.com>

changab committed:
edk II C Coding Standard: Updates 4.2 and 4.3 sections

Updates 4.2 Directory names and 4.3 file names for
the guidelines of module directory and file naming.

PR: https://github.com/tianocore-docs/edk2-CCodingStandardsSpecification/pull/2/files

Signed-off-by: Abner Chang <abner.chang@amd.com>
Cc: Ray Ni <ray.ni@intel.com>
Cc: Michael D Kinney <michael.d.kinney@intel.com>
Cc: Sunil V L <sunilvl@ventanamicro.com>
Cc: Abdul Lateef Attar <abdattar@amd.com>
Cc: Leif Lindholm <quic_llindhol@quicinc.com>

Acked-by: Sunil V L <sunilvl@ventanamicro.com>
Reviewed-by: Abdul Lateef Attar<AbdulLateef.Attar@amd.com>
Reviewed-by: Ray Ni <ray.ni@intel.com>

vlsunil commented on 4_naming_conventions/42_directory_names.md:
```diff
@@ -0,0 +1,154 @@
+   - CpuDxe/                    # First implementation of CpuDxe.
+   - CpuDxe/RiscV64/            # RiscV64 specific implementation.
+           /                    # Common files for the RiscV64 and other archs.
+   - CpuDxeRiscv64/             # RiscV64 implementation of CpuDxe.
```
In which case this will be required? Isn't generic module and arch specific folder inside sufficient always? For ex: Even if the module is created fresh for RISC-V, can we not name it generic since we don't know in future some other architecture may need it?

changab replied:
Those are all examples we can take for the module. Usually, if the module is only for the specific arch for sure then we can choose <Feature>Riscv64 (e,g, PlicDxeRiscV64). You can have CpuDxe/RiscV64 if you consider this module would have other archs. Or just use the generic name like CpuDxe. However, other archs may need a reconstruction later.  Conversations between contributor and maintainer may be required for the final decision.
Please also check the Wiki page, I added you as the collaborator,
[The Guidelines of Reconstruct EDK II Modules for Processor Architectures and Vendors' Implementation](https://github.com/changab/tianocore.github.io/wiki/The-Guidelines-of-Reconstruct-EDK-II-Modules-for-Processor-Architectures-and-Vendors'-Implementation)

vlsunil replied:
Okay. Thanks!

vlsunil commented on 4_naming_conventions/43_file_names.md:
```diff
@@ -0,0 +1,212 @@
+
+Example:
+   - SmmAmdSmmCpuFeaturesLib.inf
+   - SmmIntelSmmCpuFeaturesLib.inf
```
I still think this should be SmmAmdCpuFeaturesLib.inf. Since we can't change existing SmmCpuFeaturesLib.inf and CpuCommonFeaturesLib.inf, they can be considered as Intel related if different from AMD.

changab replied:
What we define here is generic and mainly for the new module that is introduced to edk2
From the naming guidelines perspective, SmmAmdSmmCpuFeatureLib is compliant with <phase><CpuArch><Vendor><LibraryClassName><Dependency> file naming structure. So both SmmAmdSmmCpuFeaturesLib and SmmIntelSmmCpuFeaturesLib are correct.
However, SmmCpuFeaturesLib was implemented before the guidelines, the way to reconstruct the existing module is mentioned here, [The Guidelines of Reconstruct EDK II Modules for Processor Architectures and Vendors' Implementation](https://github.com/changab/tianocore.github.io/wiki/The-Guidelines-of-Reconstruct-EDK-II-Modules-for-Processor-Architectures-and-Vendors'-Implementation).
Take SmmCpuFeaturesLib as an example, we won't touch the existing INF file. However, the newly added INF file will follow the guidelines. I will remove SmmIntelSmmCpuFeaturesLib as I know that is confusing, but I will keep AMD one because that is a new file.

vlsunil replied:
Okay.

vlsunil commented on 4_naming_conventions/43_file_names.md:
```diff
@@ -0,0 +1,212 @@
+   IA32/SmiException.nasm
+   X64/SmiException.nasm
+   IA32_X64/IntelSmmCpuFeaturesLib.c
+   IA32_X64/AmdSmmCpuFeaturesLib.c
```
Is it Ia32 or IA32? Both are used in different places.

changab replied:
Thanks for catching this. Should be Ia32

vlsunil commented on 4_naming_conventions/43_file_names.md:
```diff
@@ -58,3 +59,154 @@ regular expression:
+   IA32/SmiException.nasm
+   X64/SmiException.nasm
+   IA32_X64/IntelSmmCpuFeaturesLib.c
+   IA32_X64/AmdSmmCpuFeaturesLib.c
```
Did you modify the directory path from IA32_X64 to Ia32_X64 in the examples above? For some reason, I still see IA32.

changab replied:
Just pushed!

changab replied:
Fix at commit bcb249b1788d7b11b182e2c2e1043fc7793a071a

niruiyu commented on 4_naming_conventions/42_directory_names.md:
````diff
@@ -0,0 +1,154 @@
+
+* If the implementation is for the specific CPU architectures, CPU vendors or the implementation is shared by certain CPU archs:
+```
+<Feature><Phase><CpuArch><Vendor>
````
Below rule
\<Feature\> \<Phase\> \[ \<CpuArch\> \[ \<Vendor\> \] \]
can cover below two rules:
\<Feature\> \<Phase\> and \<Feature\>\<Phase\>\<CpuArch\>\<Vendor\>

And it also restrict that Vendor only can be used when CpuArch exists.

changab replied:
In order to have a consistent syntax in sections 4.2 and 4.3, I will use  "[]" for the optional element.

changab replied:
will fix this in the next version.

niruiyu commented on 4_naming_conventions/42_directory_names.md:
````diff
@@ -0,0 +1,154 @@
+MdeModulePkg/Universal/PCD), and the implementation is general for all CPU architectures without the specific arch dependency:
+
+```
+<Feature>/<Phase>
````
PcdDxe driver can either be in Pcd/Dxe or PcdDxe folder.
I cannot see a clear reason why we use Pcd/Dxe instead of PcdDxe.
So, I think if we define the rule in this doc, Pcd/Dxe violates the rule. But that's ok IMO.

niruiyu replied:
What I mean is we could remove the rule "\<Feature\> / \<Phase\>".

changab replied:
This rule is defined specifically for the module which has no code shared between phases. PcdDxe follows the rule \<Feature\>\<Phase\>[\<CpuArch\>[\<Vendor\>]] which is different with \<Feature\>/\<Phase\>. But we can have \<Feature\>/\<Phase\>[/\<CpuArch\>[/\<Vendor\>]] for both \<Feature\>/\<Phase\> and \<Feature\>/\<Phase\>/\<CpuArch\>/\<Vendor\>.

changab replied:
> What I mean is we could remove the rule "<Feature> / <Phase>".

Or do you prefer to have something like PcdDxe and PcdPei, no need Pcd/ directory for the different phases?

niruiyu commented on 4_naming_conventions/42_directory_names.md:
````diff
@@ -0,0 +1,154 @@
+```
+<Feature><Phase><CpuArch><Vendor>
+  or
+<Feature><Phase>/<CpuArch>/<Vendor>
````
We mixed the module path rule and the directory layout rule inside a module.
IMO, we needn't mention this rule here.

niruiyu commented on 4_naming_conventions/42_directory_names.md:
````diff
@@ -0,0 +1,154 @@
+* If the implementation does not have any shared code between phases (e.g.
+MdeModulePkg/Universal/PCD), and is for the specific CPU architectures, CPU vendors or the implementation is shared by certain CPU archs:
+```
+<Feature>/<Phase>/<CpuArch>/<Vendor>
````
No need of this rule.

changab replied:
Same as the above answer, this is a way to centralize different phase supports in one directory.
Pcd/Dxe/ is for the module that doesn't share code with PEI phase. Additionally, apply CpuArch>/<Vendor> on it.
Or do you prefer to have something like PcdDxe and PcdPei, no need Pcd/ directory for the different phases?

niruiyu commented on 4_naming_conventions/42_directory_names.md:
````diff
@@ -0,0 +1,154 @@
+```
+<Phase><CpuArch><Vendor><LibraryClassName><Dependency>
+  or
+<Phase><LibraryClassName><Dependency>/<CpuArch><Vendor>
````
Similar concern: the rule after "or" mixed the module path rule and the directory layout rule inside a module.
We could remove the rule after "or".

changab replied:
@niruiyu what do you mean here? Those two are the possible directories for the library module.

   - SmmIa32X64AmdSmmCpuFeaturesLib/
   - SmmArmAArch64SmmCpuFeaturesLib/
   - BaseMpInitLib/RiscV64/ 
                  /Ia32X64/ 
                  /ArmAArch64/  
                  /LoongArch64/
                  /

niruiyu commented on 4_naming_conventions/42_directory_names.md:
```diff
@@ -0,0 +1,154 @@
+   <CpuArch>            OPTIONAL     The <CpuArch> is represented with a BNF,
+                                     <arch> ::='Ia32' | 'X64' | 'Arm' | 'AArch64' | 'RiscV64' |
+                                               'LoongArch64' | 'Ebc'   
+                                     <CpuArch> ::= <arch>['_'<arch>]*
```
Can we remove "_"?

changab replied:
I am ok with this.

niruiyu commented on 4_naming_conventions/43_file_names.md:
````diff
@@ -0,0 +1,211 @@
+   - SmmIa32_X64SmmCpuFeaturesLib.inf
+```
+
+#### 4.3.5.4 EDK II source files within a Library/Module instance
````
We'd better not define the rule for source files. Let's give freedom to developers. :)
Any change to the source file names only impacts the INF content.

changab replied:
__FileName__ gives the freedom to developers when there is no \<CpuArch\> and \<Vendor\>, however, we should define the rules when considering the archs and vendors in the source file name. But I can remove the second part because it mixes up the file and directory naming.

niruiyu commented on 4_naming_conventions/43_file_names.md:
````diff
@@ -0,0 +1,211 @@
+* If the implementation does not have any shared code between phases:
+
+```
+<Feature>/<Phase>/<Feature>.inf
````
The rule here mixed the directory rule and file name rule.
The directory rule is defined by chapter 4.2
So the file name rule here is:
\<Feature\>.inf
or
\<Feature\> \<CpuArch\> \<Vendor\>.inf

People may argue: why \<Feature\> should be part of file name because it's already part of directory path? Why \<Phase\> is not part of file name?

Can we align with Mike and propose a simple/clear/ONE rule for all?
I like the below rule you proposed:
\<Feature\> \<Phase\> \[ \<CpuArch\>\[\<Vendor\>\]\].inf

changab replied:
Phase is already mentioned in the directory (e.g. Pcd/__Dxe__) and we at least need [Feature] as the name of a file.

changab commented on 4_naming_conventions/43_file_names.md:
````diff
@@ -0,0 +1,211 @@
+
+#### 4.4.5.3 EDK II INF file within a Library instance
+```
+<Phase><CpuArch><Vendor><LibraryClassName><Dependency>.inf
````
@niruiyu
In order to have a consistent naming rule for \<CpuArch\>\<Vendor\>, I will apply [\<CpuArch\>[\<Vendor\>]] on this case as well. A long file name for \<CpuArch\>\<Vendor\> is expected.

changab replied:
> @niruiyu
> In order to have a consistent naming rule for <CpuArch><Vendor>, I will apply [<CpuArch>[<Vendor>]] on this case as well. A long file name for <CpuArch><Vendor> is expected.

@niruiyu I still use [\<CpuArch\>][\<Vendor\>], so Vendor is unnecessary to have a dependence on CpuArch. The reason is the module is expected for X86 only, that is no need to have Ia32X64 for the vendor INF under X86 module. This concept is applied to the entire File name section. Please check the next version and see what else we can improve.

changab replied:
> Please check the next version and see what else we can improve.

PR is updated.
